spec/std/crystal/hasher_spec.cr

```diff
@@ -3,145 +3,231 @@ require "bit_array"
-      @v.should_not eq(other.@v)
-      (@v ^ other.@v).popcount.should be_close(32, 24)
-    end
+  def initialize(a, b : UInt64)
+    @hasher = FunnyHash64.new(a, b)
   end
 
-  def self.bytes(b)
-    SpecResult.new(new.bytes(b).result)
+  def self.for_test
+    new(1_u64, 1_u64)
   end
 end
 
-# Note: this tests are probabilistic.
-# Hasher is randomly seeded, therefore this tests could randomly
-# fail. Probability of fail is very very low, but not zero.
-# If test fails two times in a row, then error is certainly real.
-describe "Hasher" do
-  describe "bytes" do
-    it "should differ for every single bit and length" do
-      empty_hash = Crystal::Hasher.bytes(Bytes.new(0))
-      1.upto(64) do |byte_len|
-        bytes = Bytes.new(byte_len, 0_u8)
-        zero_hash = Crystal::Hasher.bytes(bytes)
-        prev_hash = zero_hash
-        prev_prev_hash = Crystal::Hasher::SpecResult.new
-        bit_len = byte_len * 8
-        0.upto(bit_len - 1) do |bit|
-          bytes[bit/8] ^= 1 << (bit & 7)
-          cur_hash = Crystal::Hasher.bytes(bytes)
-          cur_hash.differs empty_hash
-          cur_hash.differs zero_hash
-          cur_hash.differs prev_hash
-          cur_hash.differs prev_prev_hash
-          prev_hash, prev_prev_hash = cur_hash, prev_hash
-          # check for other random bit
-          while true
-            other_bit = rand(bit_len)
-            break if other_bit != bit
-          end
-          bytes[other_bit/8] ^= 1 << (other_bit & 7)
-          oth_hash = Crystal::Hasher.bytes(bytes)
-          cur_hash.differs oth_hash
-          bytes[other_bit/8] ^= 1 << (other_bit & 7)
-          # check for length extention
-          (byte_len - 1).downto((bit + 7)/8) do |len|
-            short_hash = Crystal::Hasher.bytes(bytes[0, len])
-            cur_hash.differs short_hash
-          end
-          bytes[bit/8] ^= 1 << (bit & 7)
-        end
+enum TestHasherEnum
+  A
+  B
+end
+
+alias THasher = Crystal::Hasher
+
+describe "Crystal::Hasher" do
+  context "behavior" do
+    it "#nil should change hasher state" do
+      hasher = THasher.for_test
+      hasher1 = nil.hash(hasher)
+      hasher2 = nil.hash(hasher1)
+      hasher1.result.should_not eq(hasher.result)
+      hasher2.result.should_not eq(hasher.result)
+      hasher2.result.should_not eq(hasher1.result)
+    end
+
+    it "#bool should change state and differ" do
+      hasher = THasher.for_test
+      hasher_true = true.hash(hasher)
+      hasher_false = false.hash(hasher)
+      hasher.result.should_not eq(hasher_true.result)
+      hasher.result.should_not eq(hasher_false.result)
+      hasher_true.result.should_not eq(hasher_false.result)
+    end
+
+    it "#int should change state and differ" do
+      hasher = THasher.for_test
+      hasher1 = 1.hash(hasher)
+      hasher2 = 2.hash(hasher)
+      hasher12 = 2.hash(hasher1)
+      [hasher, hasher1, hasher2, hasher12]
+        .map(&.result)
+        .combinations(2)
+        .each do |(a, b)|
```

Comment:
wrong indentation

Reply:
You you sure, @Sija?

```
➜  crystal git:(althasher) ✗ bin/crystal tool format src spec
Using compiled compiler at `.build/crystal'
➜  crystal git:(althasher) ✗ gd
➜  crystal git:(althasher) ✗
```

Reply:
@azkhan yes I am, block body SHOULD be indented, yet it's not. Perhaps it's a formatter bug?

Reply:
Bad idea with `{}` blocks looks better.

Reply:
`crystal tool format` made this indentation.

Using curly braces here is semantically wrong: assertion is an action, not computation or transformation.

But if you insist, I will change to curly braces. Should I?

Reply:
@Sija looks like is a bug with the formatter:

```crystal
[1, 2, 3]
  .each do |n|
  n
end
```

Let's handle that independently of this PR. Excessive nitpicking on PR tends to tire people.

Reply:
@funny-falcon IIRC `do ... end`  vs `{ ... }` is matter of multi-line vs one-line block bodies, not the "actionable" intent, therefore I'd suggest yes, changing the tests to use curly braces - which also helps with readability - one line to scan, not 3 of 'em.

Reply:
/cc MakeNowJust

Reply:
@funny-falcon lets stay things as is. they automatically fixed with with of formatter.

Reply:
> block body SHOULD be indented, yet it's not. Perhaps it's a formatter bug?

It is indented relative to statement level. It is not a bug, just current rules. May be rule should be changed, but which way?

May be, method calls should be indented by 4 spaces?
```crystal
[hasher, hasher1, hasher2, hasher12]
    .map(&.result)
    .combinations(2)
    .each do |(a, b)|
  a.should_not eq(b)
end
```

Reply:
@luislavena IMO there are 2 separate issues here: formatter bug and bad readability of `do ... end` version of block bodies - first one should be handled in separate PR, yet the second one results in badly readable code and could be fixed, so why not do it?

Reply:
> is matter of multi-line vs one-line block bodies, not the "actionable" intent,

There are different points of view on this question. I prefer curly braces only for "one-line non-actionable".

But, ok, I've changed to curly braces.

Reply:
PLEASE, let's not discuss such small syntax details. If the specs pass it's OK. This only makes the contributor feel tired and annoyed about stuff that doesn't really matter.

There's no formatter bug, the expression start at indent 0 and the block is at indent 2.